Site Editor: Add template tab to sidebar

david-szabo97 opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description
Add template tab to settings sidebar. Resolves https://github.com/WordPress/gutenberg/issues/27470

## How has this been tested?
1. Open site editor
2. Open settings sidebar
3. Make sure there are two tabs: Template and Block
4. Make sure Template tab is selected when opening the settings sidebar with no blocks are selected
5. Make sure Block tab is selected when opening the settings sidebar when a block is selected

## Screenshots <!-- if applicable -->

https://user-images.githubusercontent.com/2256104/106467518-0e731c00-649d-11eb-9adf-25d161eba5ca.mov


## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

david-szabo97 committed:
Add template panel to sidebar

david-szabo97 commented on packages/edit-site/src/components/sidebar/template-card/index.js:
```diff
@@ -0,0 +1,37 @@
+ */
+import { useSelect } from '@wordpress/data';
+import { Icon } from '@wordpress/components';
+import { page } from '@wordpress/icons';
```
Do we have an icon representing a template?

david-szabo97 committed:
Cleanup i18n

david-szabo97 committed:
Rename function

jameskoster reviewed (changes requested):
Looking good for the most part!

The icon is slightly too small though – it should be 24x24.

<img width="293" alt="Screenshot 2021-02-01 at 15 54 00" src="https://user-images.githubusercontent.com/846565/106482844-cd333a00-64a5-11eb-982a-0d1ffd644328.png">

Also, let's use the `layout` [icon](https://github.com/WordPress/gutenberg/blob/master/packages/icons/src/library/layout.js):

<img width="496" alt="Screenshot 2021-02-01 at 15 56 37" src="https://user-images.githubusercontent.com/846565/106483119-184d4d00-64a6-11eb-8372-25997eac64b3.png">

david-szabo97 committed:
Fix icon size and change icon

david-szabo97 commented:
Fixed the size and changed the icon. This one looks different though 🤔 I used the layout
![image](https://user-images.githubusercontent.com/2256104/106515721-5a8e8280-64d6-11eb-9965-2519e77b40e2.png)

david-szabo97 committed:
Fix icon shrinking

david-szabo97 committed:
Hide while loading

jameskoster commented:
Huh! Looks like that icon didn't get updated in the code. I'll handle that in a separate PR.

I noticed the spacing and alignment doesn't match the icon/name/description on the block tab:

<img width="458" alt="Screenshot 2021-02-02 at 09 55 22" src="https://user-images.githubusercontent.com/846565/106583319-e0ddaf80-653c-11eb-95cb-79bc10f67e7f.png">

If we can tidy that up then I think we are good to go on the design side :)

david-szabo97 commented:
The funny thing is, I actually copied the styles of that layout. But there is a bug in that one that causes the icon to shrink if the description is long:

https://user-images.githubusercontent.com/2256104/106607830-08943e00-6564-11eb-8057-d3aeb033a5db.mov

Here is a recording with `flex: 0 0 $button-size` applied. (ie. icon can't grow or shrink, stays fixed at $button-size)

https://user-images.githubusercontent.com/2256104/106607902-1cd83b00-6564-11eb-9b5d-ae9b075f1c2c.mov

david-szabo97 committed:
Fix shrinking icon

david-szabo97 commented:
Since the block card's styling is unrelated to this PR, I'll spin up a new PR for that one.

david-szabo97 commented:
Rebased to update the icon.

jeyip commented:
## Testing
### Behavior
- [x] Template tab is selected when opening the settings sidebar with no blocks are selected
- [x] Block tab is selected when opening the settings sidebar when a block is selected 

### Browsers
- [x] Chrome
- [x] Firefox
- [x] Edge
- [x] Safari

youknowriad commented:
It seems we're getting closer to briding edit-post and edit-site. Opening the template in the post editor already shows a "template" tab on the sidebar. I believe  at some point edit-site could become just a proxy to edit-post with a bunch of extra settings...

david-szabo97 committed:
Rename toggle

david-szabo97 committed:
Add e2e tests

david-szabo97 committed:
Remove warning and refactor one test

david-szabo97 committed:
Replace store strings with definitions

Copons commented on packages/edit-site/src/components/sidebar/header/index.js:
```diff
@@ -0,0 +1,71 @@
+import { STORE_NAME } from '../../../store/constants';
+import { SIDEBAR_BLOCK, SIDEBAR_TEMPLATE } from '../constants';
+
+const SettingsHeader = ( { sidebarName } ) => {
```
Nitpick: what about calling this `SidebarHeader`?
I was thrown off by the difference between component name and file name (`sidebar/header`).

In the same vein, it would also be fine to rename the folder to `sidebar/settings-header` (but imho it's less clear).

david-szabo97 replied:
I renamed the folder to `settings-heade`. `edit-post` is using `sidebar/settings-header` as well. `SettingsSidebar` is only used together with the default (settings) sidebar (it's not used with the global styles sidebar). That's why I'd keep the `Settings` in the name.

vindl replied:
I think it makes sense to align the naming with `edit-post` here.

david-szabo97 committed:
Rename folder to settings-header

vindl reviewed (commented):
The tests are failing after the folder rename. Imports in `packages/edit-site/src/style.scss` also have to be updated.

david-szabo97 committed:
Rename

youknowriad commented:
Seems like this PR might have broken an e2e test on master. Or at least, the failure started happening with this merge.

david-szabo97 commented:
> Seems like this PR might have broken an e2e test on master. Or at least, the failure started happening with this merge.

Created a draft PR to see if that fixes the e2e tests. https://github.com/WordPress/gutenberg/pull/28950

david-szabo97 commented:
Looks like some changes were made in this PR https://github.com/WordPress/gutenberg/pull/28835 that removed the custom component based toolbar control. This PR still included that check so probably that's why it's failing.
